File: src/main/java/by/sorface/sorface/gateway/utils/JsonHttpResponseUtils.java
package by.sorface.sorface.gateway.utils;

import by.sorface.sorface.gateway.records.ErrorOperation;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.io.buffer.DataBuffer;
import org.springframework.core.io.buffer.DataBufferFactory;
import org.springframework.core.io.buffer.DataBufferUtils;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpResponse;
import reactor.core.publisher.Mono;

import java.util.Map;

public final class JsonHttpResponseUtils {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    public static Mono<Void> toJsonException(final ServerHttpResponse response, final HttpStatus httpStatus, final Throwable throwable) {
        return toJsonException(response, httpStatus, throwable.getMessage());
    }

    public static Mono<Void> toJsonException(final ServerHttpResponse response, final HttpStatus httpStatus, final String message) {
        final DataBufferFactory dataBufferFactory = response.bufferFactory();

        DataBuffer buffer;

        try {
            buffer = dataBufferFactory.wrap(OBJECT_MAPPER.writeValueAsBytes(new ErrorOperation(httpStatus.value(), message)));
        } catch (JsonProcessingException e) {
            return Mono.error(e);
        }

        return response.writeWith(Mono.just(buffer)).doOnError((error) -> DataBufferUtils.release(buffer));
    }

}

File: src/main/kotlin/by/sorface/gateway/utils/JsonHttpResponseUtils.kt
package by.sorface.gateway.utils

import by.sorface.gateway.records.ErrorOperation
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.http.HttpStatus
import org.springframework.http.server.reactive.ServerHttpResponse
import reactor.core.publisher.Mono

object JsonHttpResponseUtils {

    private val OBJECT_MAPPER = ObjectMapper()

    fun buildJsonResponseWithException(response: ServerHttpResponse, httpStatus: HttpStatus = HttpStatus.INTERNAL_SERVER_ERROR, throwable: Throwable): Mono<Void> {
        return buildJsonResponseWithException(response, httpStatus, throwable.message)
    }

    fun buildJsonResponseWithException(response: ServerHttpResponse, httpStatus: HttpStatus = HttpStatus.INTERNAL_SERVER_ERROR, message: String? = null): Mono<Void> {
        return buildJsonResponseWithException(response, httpStatus, ErrorOperation(httpStatus.value(), message))
    }

    private fun buildJsonResponseWithException(response: ServerHttpResponse, httpStatus: HttpStatus, jsonObject: Any): Mono<Void> {
        val dataBufferFactory = response.bufferFactory()

        response.setStatusCode(httpStatus)

        val dataBufferMono = Mono.defer {
            var result: Mono<ByteArray>

            try {
                val data = OBJECT_MAPPER.writeValueAsBytes(jsonObject)

                result = Mono.just(data)
            } catch (e: JsonProcessingException) {
                result = Mono.error(e)
            }

            result
        }
            .map { bytes: ByteArray -> dataBufferFactory.wrap(bytes) }

        return response.writeWith(dataBufferMono)
    }
}
